File: src/main/java/io/github/tanguygab/tabadditions/velocity/VelocityListener.java
package io.github.tanguygab.tabadditions.velocity;

import com.google.common.io.ByteArrayDataInput;
import com.google.common.io.ByteStreams;
import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.command.CommandExecuteEvent;
import com.velocitypowered.api.event.connection.PluginMessageEvent;
import com.velocitypowered.api.event.player.PlayerChatEvent;
import com.velocitypowered.api.proxy.Player;
import io.github.tanguygab.tabadditions.shared.features.chat.Chat;
import me.neznamy.tab.shared.TAB;
import me.neznamy.tab.shared.platform.TabPlayer;

public class VelocityListener {

    private final TAB tab = TAB.getInstance();

    @Subscribe
    public void onChat(PlayerChatEvent e) {
        if (!e.getResult().isAllowed()) return;
        Chat chat = tab.getFeatureManager().getFeature("Chat");
        if (chat == null || chat.isBukkitBridgeChatEnabled()) return;

        e.setResult(PlayerChatEvent.ChatResult.denied());
        chat.onChat(getPlayer(e.getPlayer()), e.getMessage());
    }

    @Subscribe
    public void onCommand(CommandExecuteEvent e) {
        if (!e.getResult().isAllowed()) return;
        Chat chat = tab.getFeatureManager().getFeature("Chat");
        if (chat == null || !(e.getCommandSource() instanceof Player sender)) return;
        TabPlayer player = getPlayer(sender);
        if (chat.onCommand(player,e.getCommand())) {
            e.setResult(CommandExecuteEvent.CommandResult.denied());
        }
    }


    @Subscribe
    public void onMessageReceived(PluginMessageEvent e) {
        if (e.getIdentifier() != VelocityPlatform.IDENTIFIER || !(e.getSource() instanceof Player receiver)) return;

        ByteArrayDataInput in = ByteStreams.newDataInput(e.getData());
        if (!in.readUTF().equalsIgnoreCase("Chat")) return;
        Chat chat = tab.getFeatureManager().getFeature("Chat");
        if (chat != null) chat.onChat(getPlayer(receiver),in.readUTF());
    }

    private TabPlayer getPlayer(Player player) {
        return tab.getPlayer(player.getUniqueId());
    }

}

File: src/main/kotlin/io/github/tanguygab/tabadditions/velocity/VelocityListener.kt
package io.github.tanguygab.tabadditions.velocity

import com.google.common.io.ByteStreams
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.connection.PluginMessageEvent
import com.velocitypowered.api.event.player.PlayerChatEvent
import com.velocitypowered.api.proxy.Player
import io.github.tanguygab.tabadditions.shared.features.chat.Chat
import me.neznamy.tab.shared.TAB

class VelocityListener {
    private val tab = TAB.getInstance()

    @Subscribe
    fun onChat(e: PlayerChatEvent) {
        if (!e.result.isAllowed) return
        val chat = tab.featureManager.getFeature<Chat>("Chat")
        if (chat == null || chat.bukkitBridgeChatEnabled) return

        e.result = PlayerChatEvent.ChatResult.denied()
        chat.onChat(getPlayer(e.player), e.message)
    }

    @Subscribe
    fun onMessageReceived(e: PluginMessageEvent) {
        val receiver = e.source
        if (e.identifier !== VelocityPlatform.IDENTIFIER || receiver !is Player) return

        val input = ByteStreams.newDataInput(e.data)
        if (!input.readUTF().equals("Chat", ignoreCase = true)) return

        val chat = tab.featureManager.getFeature<Chat>("Chat")
        chat?.onChat(getPlayer(receiver), input.readUTF())
    }

    private fun getPlayer(player: Player) = tab.getPlayer(player.uniqueId)!!
}
